WIP: e2e: node: add and revamp cpumanager tests

ffromani opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Start the cleanup of cpumanager tests and add explicit cases to cover https://github.com/kubernetes/kubernetes/issues/103952

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/100145

#### Special notes for your reviewer:
Once this lands, will send similar PRs for the other resource managers; however the relevant e2e tests are at different level of bitrot. The cpumanager tests are among the ones which need more urgent attention.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

ffromani committed:
e2e: node: add and extend cpumanager tests

Add tests to explicitly cover the issue
https://github.com/kubernetes/kubernetes/issues/103952

Take the chance to start new code layout and verification
approach which will be used later to revamp and extend
the existing tests.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/sig node

ffromani commented:
/test

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented:
/retest

ffromani commented:
/retest

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2

ffromani commented:
/retest

ffromani commented:
retesting because: https://github.com/kubernetes/test-infra/issues/30759

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2

ffromani commented:
ok, this worked fine locally but seems flaky. Looking for alternatives

swatisehgal commented:
/cc

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented:
took more iterations than expected (likely busybox image too old?) but now the foundations have been laid and the PR is reviewable

shajmakh commented on test/e2e_node/cpu_manager_test.go:
```diff
@@ -47,27 +49,130 @@ type ctnAttribute struct {
+}
+
+func (ca ctnAttribute) WithDefaults() ctnAttribute {
+	cmd := "grep Cpus_allowed_list /proc/self/status | cut -f2 && sleep 1d"
```
I wonder if the allowed cpus will still be reflected properly under the status if cgroup is of v2, If not mistaken I saw that empty, but currently I don't have a cluster with v2. I'll deploy one and check.

shajmakh commented on test/e2e_node/cpu_manager_test.go:
```diff
@@ -47,27 +49,130 @@ type ctnAttribute struct {
+	}
+}
+
+func (ca ctnAttribute) WithDefaults() ctnAttribute {
```
the naming (or the functionality maybe?) could be a bit misleading if one don't check the implementation of this function, as I would have expected it to set default values to all the elements of the struct if they aren't initialized. unless the tests here assume that the cpus are always set.

ffromani commented:
/hold

need to document why we are using `Ordered`

ffromani committed:
e2e: node:  move SMT tests to proper It blocks

Building on top of the new e2e cpumanager layout,
make the SMT alignment check become regulkar It blocks
to demonstrate how the transition will progress.

No intended change of behavior.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/hold cancel

ffromani commented:
followup PRs will continue the rewrite/extension of cpumanager tests and their de-entanglement with topology manager tests

bart0sh commented:
/triage accepted
/priority important-soon

swatisehgal commented:
/assign

swatisehgal commented on test/e2e_node/cpu_manager_test.go:
```diff
@@ -751,12 +856,251 @@ func isSMTAlignmentError(pod *v1.Pod) bool {
+					if !ok {
+						return fmt.Errorf("local node not ready yet")
+					}
+					allocatableCPUQty := node.Status.Allocatable[v1.ResourceCPU]
+					allocatableCPU := allocatableCPUQty.Value()
```
We populated `allocatableCPU` using `getLocalNodeCPUDetails` with allocatable CPUs on the nodes in the BeforeEach ([here](https://github.com/kubernetes/kubernetes/pull/120754/commits/a23f3aec439005fe5a67a2e52061a3f67ae04a5e#diff-9f37461eb6db1067ff11d6118e4ea593c117772b66cb12d4725e52f9419384ecR980)). Why don't we reuse that rather than getting that again from node status?

Also, we have a case of variable shadowing here.

swatisehgal commented on test/e2e_node/cpu_manager_test.go:
```diff
@@ -906,6 +796,149 @@ var _ = SIGDescribe("CPU Manager [Serial] [Feature:CPUManager]", func() {
+				// positive test: try to run a container whose requests are a multiple of SMT level, check allocated cores
+				// 1. are core siblings
+				// 2. take a full core
+				// WARNING: this assumes 2-way SMT systems - we don't know how to access other SMT levels.
```
I wonder if it would be better for us to determine CPUs per core on the system and then use that when we specify the number of CPUs in the pod CPU request? It should be straightforward and would allow us to get rid of the assumption that the system has 2 CPUs on a core (i.e. 2 way SMT).

swatisehgal reviewed (commented):
Thanks for your work on this and apologies for the delay in getting to this.

This PR is a significant improvement in the CPU Manager tests and gets rid of a tonne of assumptions. I left a few comments but overall the PR is looking good!

ffromani commented:
/test pull-crio-cgroupv1-node-e2e-resource-managers
/test pull-kubernetes-node-e2e-containerd-ec2
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-kubetest2
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2
/test pull-kubernetes-node-kubelet-serial-topology-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager-kubetest2

ffromani commented:
moving back to WIP: in addition to review comments, need to rebase
